### PR TITLE
Generate GLTF.Meshes as UnityEngine.Meshes instead of GameObjects

### LINKED ISSUE #370: Do not create GameObjects for mesh primitives
Instead, represent multiple mesh primitives as Unity submeshes. So instead of the post-load scene graph being like this:

* gltf_node => GameObject
    * gltf_mesh => GameObject
        * gltf_primitive => GameObject
            * mesh_data + gltf_material => Renderer
        * gltf_primitive => GameObject
            * mesh_data + gltf_material => Renderer

...it would look like this:

* gltf_node => GameObject
    * gltf_mesh_primitive[] + gltf_material[] => Renderer

See the documentation for [Mesh.SetTriangles](https://docs.unity3d.com/ScriptReference/Mesh.SetTriangles.html) for the backing structure.

### PR DESCRIPTION
**Pull request by stevenvergenz**
Fixes #370

### CODE REVIEW

File: UnityGLTF/Assets/UnityGLTF/Scripts/GLTFSceneImporter.cs [hunk 1]
```diff
@@ -1529,230 +1543,171 @@ private void NormalizeBoneWeightArray(Vector4[] weights)
...
-				(!shouldUseDefaultMaterial && _assetCache.MaterialCache[materialIndex] == null))
-			{
-				await ConstructMaterial(materialToLoad, shouldUseDefaultMaterial ? -1 : materialIndex);
-			}
+			await ConstructUnityMesh(unityData, meshIndex, mesh.Name);
 		}
 
-		protected UnityMeshData ConvertAccessorsToUnityTypes(MeshConstructionData meshConstructionData)
+		protected void ConvertAttributeAccessorsToUnityTypes(
+			ref MeshCacheData.PrimitiveCacheData primData,
```
Comment (AdamMitchell-ms) [inline UnityGLTF/Assets/UnityGLTF/Scripts/GLTFSceneImporter.cs]: Neither of these need to be passed in by ref.

File: UnityGLTF/Assets/UnityGLTF/Scripts/GLTFSceneImporter.cs [hunk 2]
```diff
@@ -1529,230 +1543,171 @@ private void NormalizeBoneWeightArray(Vector4[] weights)
...
+					meshAttributes[SemanticProperties.Weight[0]].AccessorContent.AsVec4s.ToUnityVector4Raw(),
+					ref unityData.BoneWeights,
+					vertOffset);
+			}
 
-				Tangents = primitive.Attributes.ContainsKey(SemanticProperties.TANGENT)
-					? meshAttributes[SemanticProperties.TANGENT].AccessorContent.AsTangents.ToUnityVector4Raw()
-					: null,
+			if (meshAttributes.ContainsKey(SemanticProperties.POSITION))
+				meshAttributes[SemanticProperties.POSITION].AccessorContent.AsVertices.ToUnityVector3Raw(unityData.Vertices, vertOffset);
```
Comment (AdamMitchell-ms) [inline UnityGLTF/Assets/UnityGLTF/Scripts/GLTFSceneImporter.cs]: nit: braces

File: GLTFSerialization/GLTFSerialization/GLTFHelpers.cs [hunk 3]
```diff
@@ -941,5 +878,57 @@ private static string UpdateCanonicalPath(string oldPath, string newCanonicalPat
...
+		public static NodeId FindCommonAncestor(IEnumerable<NodeId> nodes)
+		{
+			// build parentage
+			GLTFRoot root = nodes.First().Root;
+			Dictionary<int, int> parentage = new Dictionary<int, int>(root.Nodes.Count);
+			for (int i = 0; i < root.Nodes.Count; i++)
+			{
+				if (root.Nodes[i].Children == null) continue;
+
+				foreach (NodeId j in root.Nodes[i].Children)
```
Comment (AdamMitchell-ms) [inline GLTFSerialization/GLTFSerialization/GLTFHelpers.cs]: I'd rename "j" to "childNode" or "child"

File: GLTFSerialization/GLTFSerialization/GLTFHelpers.cs [hunk 4]
```diff
@@ -26,332 +27,268 @@ private struct PreviousGLTFSizes
...
+						break;
+					default:
+						throw new System.Exception($"Unrecognized mesh attribute [{kvp.Key}]");
+				}
+				
 				attributeAccessor.AccessorContent = resultArray;
 			}
 		}
 
 		public static void BuildTargetAttributes (ref Dictionary<string, AttributeAccessor> attributes)
```
[the remark this answers is not in the document]
Comment (stevenvergenz) [reply GLTFSerialization/GLTFSerialization/GLTFHelpers.cs]: I think the goal was to minimize boxing? Don't know the rationale.

---
In reply to: [307931739](https://github.com/KhronosGroup/UnityGLTF/pull/470#discussion_r307931739) [](ancestors = 307931739)

File: GLTFSerialization/GLTFSerialization/GLTFHelpers.cs [hunk 5]
```diff
@@ -26,332 +27,268 @@ private struct PreviousGLTFSizes
...
-			}
-		}
-
-		/// <summary>
-		/// Uses the accessor to parse the buffer into attributes needed to construct the mesh primitive
-		/// </summary>
-		/// <param name="attributes">A dictionary that contains a mapping of attribute name to data needed to parse</param>
-		/// <summary>
-		/// Uses the accessor to parse the buffer into attributes needed to construct the mesh primitive
+		/// Removes references to indexes that do not exist.
```
[the remark this answers is not in the document]
Comment (stevenvergenz) [reply GLTFSerialization/GLTFSerialization/GLTFHelpers.cs]: Yes, VS was complaining so I standardized on windows line endings.

---
In reply to: [307932331](https://github.com/KhronosGroup/UnityGLTF/pull/470#discussion_r307932331) [](ancestors = 307932331)

**Pull request merged**